File: 25-reverse-nodes-in-k-group/25-reverse-nodes-in-k-group.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    
    public void rev(ListNode start, ListNode end) {
        ListNode prev = null, curr = start, next = start.next;
        
        while(prev != end) {
            curr.next = prev;
            prev = curr;
            curr = next;
            if(next != null) next = next.next; 
        }
    }
    
    public ListNode reverseKGroup(ListNode head, int k) {
        
        if(k == 1 || head == null || head.next == null) return head;
        ListNode dummy = new ListNode(-1);
        dummy.next = head;
        ListNode beforeStart = dummy, end = head; // beforeStart is used to point previous of start
        int i = 0;
        
        while(end != null) {
            i++;
            if(i%k == 0) { // if i modulus k gives 0 that means we can make a group with k nodes
                ListNode start = beforeStart.next, afterEnd = end.next; // afetrEnd is for next of end
                rev(start, end);
                beforeStart.next = end; // after reversing we point beforeStart to end which is the start node of reversed group
                start.next = afterEnd; // and point start to afterEnd as start is the last node of reversed group
                beforeStart = start; // now beforeStart points to start so that we can deduce the new start therefore
                end = afterEnd;
            } else end = end.next;
        }
        
        return dummy.next;
    }
}

File: 25-reverse-nodes-in-k-group/25-reverse-nodes-in-k-group.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    
    void rev(ListNode *start, ListNode *end) {
        ListNode *prev = NULL, *curr = start, *next = start->next;
        
        while(prev != end) { // when prev reaches end curr gets out of boundary so we break the loop
            curr->next = prev;
            prev = curr;
            curr = next;
            if(next != NULL) next = next->next;
        }
    }
    
    ListNode* reverseKGroup(ListNode* head, int k) {
        
        if(k == 1 || head == NULL || head->next == NULL) return head;
        
        ListNode *start = head, *end = head;
        int len = k - 1;
        
        while(len--) {
            end = end->next;
            if(end == NULL) return head; // if end pointer reaches null then we return head as it'll not form k-group
        }
        
        ListNode *nextHead = reverseKGroup(end->next, k); // to get the head of node which is failing to form k-group
        rev(start, end); // reversing k-groups
        start->next = nextHead; // merging prev k-group with the head of next group
        return end; // we return end as it'll point to the reversed head
    }
};